LLVM 13 and 14

rdeterre opened the pull request:
Specify library name and version:  **llvm/13.0.0**

This is a new PR for the changes from @Manu343726 and @tcanabrava to add an LLVM recipe.

Manu343726 committed:
llvm 10.0.0 recipe (wip)

Manu343726 committed:
Restrict Vistual Studio version

Manu343726 committed:
Add fPIC option

Manu343726 committed:
Ignore share folder

Manu343726 committed:
Add llvm license file to package

Manu343726 committed:
Fix enabling of llvm subprojects

Manu343726 committed:
Fix license copy and package ignore logic

tcanabrava committed:
update package to llvm 12.0.1 from 10.0.0

tcanabrava committed:
Check for MSVC 2019

tcanabrava committed:
Remove verbose makefile for test project

tcanabrava committed:
Bump minimal cmake version on test project

tcanabrava committed:
Extract cmake_configure into a method, and call it on install and package

tcanabrava committed:
Default to no bindings as those are not userful on a c++ project

tcanabrava committed:
Check that the compiler version is compatible with the software

tcanabrava committed:
Move the MSVC check to the validate method.

rdeterre committed:
Update to LLVM 13.0.0

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=7613) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=7613) before we can accept your contribution.<br/>**0** out of **4** committers have signed the CLA.<br/><br/>:x: Manu343726<br/>:x: tcanabrava<br/>:x: rdeterre<br/>:x: alazartech<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/conan-io/conan-center-index?pullRequest=7613) it.</sub>

rdeterre committed:
Dummy change to trigger re-build

rdeterre committed:
Restart build

rdeterre committed:
Remove GCC 6 for now, which was giving an internal compiler error

tcanabrava commented:
Thanks for this, I see that there are CI errors (I was not ever able to remove the CI errors on llvm 12)
if you need anything from me, please shout.

rdeterre committed:
Restart build

rdeterre committed:
Limit GCC to 8+

It looks like continuous integration fails because of a timeout problem. To test
this, I'm limitting the configurations that are being built

rdeterre committed:
Limit GCC to 10+

rdeterre committed:
Dummy change to trigger re-build

alazartech committed:
Limit parallel linking jobs

rdeterre committed:
Use Ninja generator

rdeterre committed:
Limit compile jobs to 4

rdeterre committed:
Limit compile jobs to 2

rdeterre committed:
Limit compile jobs to 1...

rdeterre committed:
Dummy change to trigger re-build

rdeterre commented:
Are the builds already running on the large cluster? I'm not sure what else I can try to lower the memory consumption.

rdeterre committed:
Non-parallel make build

danimtb commented:
@rdeterre yes, we enabled llvm to use the large cluster for linux builds but I think the resource requirements are higher. We are studying adding a new level of memory resources for this kind of builds but would like to make sure we have done anything possible to reduce the build requirements as thif may affect anyone building this recipe from sources

rdeterre committed:
Remove ninja dependency

SSE4 commented:
seems like still getting `collect2: fatal error: ld terminated with signal 9 [Killed]` even on `xlarge` pool :(

rdeterre commented:
Strange. I tried this on a test machine here, and I see about 2 GB memory usage.

Since everything is single-threaded now, the build takes much longer, almost 3 hours on my machine.

![image](https://user-images.githubusercontent.com/3432890/141830168-5df414ee-e0ef-4d48-a5b3-971b780de495.png)

rdeterre commented:
Is there anything else than memory usage or a timeout that could cause the failures we see? I'm a little lost to know what is wrong.

samuel-emrys commented:
My build replicates the results above with respect to memory footprint. I see memory usage in the order of ~1.5GB
![llvm_13_build](https://user-images.githubusercontent.com/36361864/150665010-840e138b-f9bd-4849-8d34-91b96fd9df5f.png)

I don't think this is a memory issue. Can someone trigger a rebuild (opening/closing the issue is usually sufficient to do this)? The build logs have expired/been cleared from the CI details page.

rdeterre commented:
Thanks for testing it out. I'll close/reopen the PR right away

rdeterre committed:
Dummy change to trigger re-build

samuel-emrys commented:
While waiting for the CI tasks to finish, I also profiled a parallel build using gcc 10.3 - substantial build time reduction and memory footprint increase as noted previously:
![llvm_13_build_parallel](https://user-images.githubusercontent.com/36361864/150669259-2608df57-15e6-4c97-9ad5-a81fca5e3f2a.png)

Instead of hardcoding `cmake = CMake(self, parallel=False);` into the recipe, perhaps it would be worth making parallel builds an option on the recipe that defaults to `False` for the CI's benefit?

uilianries commented on recipes/llvm/all/conanfile.py:
```diff
@@ -0,0 +1,125 @@
+    description = 'The LLVM Project is a collection of modular and reusable compiler and toolchain technologies'
+    url = 'https://github.com/conan-io/conan-center-index'
+    homepage = 'https://github.com/llvm/llvm-project'
+    license = 'Apache 2.0'
```
```suggestion
    license = 'Apache-2.0'
```

https://spdx.org/licenses/Apache-2.0

rdeterre committed:
Update recipes/llvm/all/conanfile.py

Co-authored-by: Uilian Ries <uilianries@gmail.com>

rdeterre commented:
> Instead of hardcoding `cmake = CMake(self, parallel=False);` into the recipe, perhaps it would be worth making parallel builds an option on the recipe that defaults to `False` for the CI's benefit? I can imagine that if anybody wants to build this themselves it would be a bit of a nightmare constrained to a single core.

Yes, this makes sense. I hardcoded  `parallel=False` for now to be 100% sure that memory isn't the issue. When the build passes, it would be great to have a solution to make it selectable.

rdeterre commented on recipes/llvm/all/conanfile.py:
```diff
@@ -0,0 +1,125 @@
+
+    def package_info(self):
+        self.cpp_info.libs = tools.collect_libs(self)
+        self.cpp_info.builddirs = ['lib/cmake']
```
I'm not 100% sure, as I don't have the test machine I was using to build this available. I'll report back when it is.

rdeterre replied:
Yes, the compiler is packaged

rdeterre committed:
Update recipes/llvm/all/conanfile.py

Co-authored-by: Uilian Ries <uilianries@gmail.com>

SSE4 commented:
what I can see on CI is process killed at `Maximum resident set size (kbytes): 8944752` (~8.7GiB), but it's still far from the limit (16GiB) for xlarge pool.

rdeterre commented:
That's good to know, it confirms that memory is not the issue here. 

Could it be just a timeout, or maybe a disk space limitation? The "build" directory on my test machine weighs 4.4 GB once everything is done.

SSE4 commented:
> That's good to know, it confirms that memory is not the issue here.
> 
> Could it be just a timeout, or maybe a disk space limitation? The "build" directory on my test machine weighs 4.4 GB once everything is done.

unfortunately, I don't know much about how these builds are internally organized. I know they are using k8s, and just pass memory limits to it. I am not k8s expert, and it will take me some time to get familiar with it and try to replicate a problem locally. in theory, if it uses some RAM drive for the disk, it could explain that part of RAM could be occupied.
but usually, if the build fails due to disk space, the errors are different. my theory is it's an OOM killer, but I don't know much how OOM killer works together with k8s, and how to obtain some useful logs with it. I'll keep you posted once I find out more.

rdeterre committed:
Support both libstdc++ ABIs

rdeterre committed:
That didn't work

tcanabrava commented:
Should we wait for llvm14 release in a few weeks and try again from there?

rdeterre commented:
No problem, I will update this merge request when LLVM 14 is released. I'm not sure if the build will pass though, it feels like we're hitting a limit on the CI machines.

rdeterre committed:
Use LLD linker

rdeterre commented:
While building this recipe locally on a machine with 16 GB of RAM, I see the same `ld killed with signal 9` error for debug builds, no problem in release mode. This happens even with `parallel=False`, which is consistent with what we see on the CI machines. I'm not sure why I wasn't seeing large RAM use when building on my other machines that have more RAM, but at least I can see the problem locally now.

One common recommendation seems to be to switch GNU LD for another linker, LLD or gold. I just pushed a commit to try using LLD.

StellaSmith commented:
I have a few suggestions:
- Use the `-l N` Ninja arguments: this tells Ninja to not spawn more processes when average system load surpasses `N`, i tried compiling using `conans.client.tools.oss.CpuProperties.get_cpus() * 0.75`
    ```py
    max_load = cpu_count * .75
    cmake.build(args=["--", "-l{}".format(max_load)])
    ```
- Maybe disable debug build: at least in the cci. In my system in which i have 16GiB of ram and 16GiB swap partition in an nvme drive, when linking, memory usage skyrockets, and while it seems to be making some progress (although really slow) i don't think it's that worth it. The system load also skyrockets due to the system moving memory in and out of swap every time.
    So if anyone really needs a debug build, they better have a good system to build it.
- Install the stripped binaries for release: This can be done by use the target `install/strip` instead of what `cmake.install()` does (which is basically a `cmake --build build_subfolder --target install`), so instead it would need to use `cmake.build(target="install/strip")`

jusito commented:
Same suggestion, just slightly different. If the issues are to many jobs at the same time but limiting it to a fix amount causes issues on other ends, why not calculate job pools according to current free resources?

Consuming the llvm/12.0.0 and link static against it has the same issue for us. ld used around 3-4GB of ram per job, so we read physical free memory before it and than calculated the amount of link/compile jobs. Because link and compile jobs can run at the same time, link jobs -1 or -2.

snippet:
```cmake
# after coverage include!
# limit linking jobs because of heavy resource usage
option(LOW_RESOURCES "enforce very low resource usage, shouldn't be needed but a quick workaround which will limit jobs to 1 linking / 2 compilation jobs" OFF)
message(STATUS "Low resource option: " ${LOW_RESOURCES})
cmake_host_system_information(RESULT system_ram QUERY AVAILABLE_PHYSICAL_MEMORY)
# linking / compiling is done at the same time and to much memory consumption will break ci / slow done it a lot
math(EXPR system_ram_possible_jobs "${system_ram} / 4000 - 2" OUTPUT_FORMAT DECIMAL)
if (system_ram_possible_jobs LESS 1 OR LOW_RESOURCES)
    set(system_ram_possible_jobs 1)
endif()
cmake_host_system_information(RESULT system_cores QUERY NUMBER_OF_LOGICAL_CORES)
get_property(TEMP_JOB_POOLS GLOBAL PROPERTY JOB_POOLS)
if (system_ram_possible_jobs LESS system_cores)
    math(EXPR compile_jobs "${system_ram_possible_jobs} * 2" OUTPUT_FORMAT DECIMAL)
    message("Your System is Memory limited, linking jobs: ${system_ram_possible_jobs} compile jobs: ${compile_jobs}")
    set_property(GLOBAL PROPERTY JOB_POOLS ${TEMP_JOB_POOLS} linking=${system_ram_possible_jobs} compile=${compile_jobs})
else()
    math(EXPR compile_jobs "${system_cores} * 1" OUTPUT_FORMAT DECIMAL)
    message("Your System is CPU limited, linking jobs: ${system_cores} compile jobs: ${compile_jobs}")
    set_property(GLOBAL PROPERTY JOB_POOLS ${TEMP_JOB_POOLS} linking=${system_cores} compile=${compile_jobs})
endif()
unset(TEMP_JOB_POOLS)

if (NOT CMAKE_JOB_POOL_LINK)
    set(CMAKE_JOB_POOL_LINK linking)
else()
    message("Using provided job pool list for linking: ${CMAKE_JOB_POOL_LINK}")
endif()
if (NOT CMAKE_JOB_POOL_COMPILE)
    set(CMAKE_JOB_POOL_COMPILE compile)
else()
    message("Using provided job pool list for linking: ${CMAKE_JOB_POOL_COMPILE}")
endif()
```

rdeterre commented:
Thanks for the feedback @StellaSmith and @jusito. I originally forced using Ninja for the build and saw even worse memory usage, but it's certainly worth trying again. For now, I pushed a change which should disable Debug builds, just to see if I can get the build to pass.

Dynamically changing the level of parallelism based on available memory really sounds like the way to go. Thanks for the code snipped, I'll see if I can put that in here.

rdeterre committed:
Disable debug builds

TheClonerx commented:
Wouldn't an option defaulted to false be better? This would require the consumer of a debug build explicitly allow it, instead of just totally disallowing it.

canatella commented:
@TheClonerx :

> Wouldn't an option defaulted to false be better? This would require the consumer of a debug build explicitly allow it, instead of just totally disallowing it.

My 2 cents: Correct me if I'm wrong but LLVM debug builds are only useful for LLVM developers which are not using Conan to build LLVM anyway. It does not change anything to the code produced by LLVM, so what would be the point of an option?  If you add an option people will enable it and will complain that it's not working when they don't have a powerful enough machine to build it.

samuel-emrys commented:
> @TheClonerx :
> 
> > Wouldn't an option defaulted to false be better? This would require the consumer of a debug build explicitly allow it, instead of just totally disallowing it.
> 
> My 2 cents: Correct me if I'm wrong but LLVM debug builds are only useful for LLVM developers which are not using Conan to build LLVM anyway. It does not change anything to the code produced by LLVM, so what would be the point of an option? If you add an option people will enable it and will complain that it's not working when they don't have a powerful enough machine to build it.

I don't think the package manager should inject any assumptions about how a package is going to be used here. It should provide all of the build configurations possible upstream. A warning may be appropriate to inform consumers of the build requirements, or maybe this is better placed in the markdown generator output (is there a way to modify this?)

Regardless, I don't think protecting users from themselves in this way is an appropriate action to take. This configuration is only required to get it to build on the CCI CI runners, so users should have the option to turn this configuration off. I think that this applies to both debug and parallel builds.

rdeterre commented:
> I don't think the package manager should inject any assumptions about how a package is going to be used here. It should provide all of the build configurations possible upstream.

Yes, that makes sense. For example, one application I'm looking at is using LibTooling and not being able to build it in debug mode is a pain.

So far, I got rid of debug builds by throwing an exception in the `validate` method just to test if the build passes (and I'm happy it does). What would be the right solution to prevent debug builds in CCI but allow them for people building the library themselves?

samuel-emrys commented:
> What would be the right solution to prevent debug builds in CCI but allow them for people building the library themselves?

I'll let one of the maintainers comment on the "right" solution, but _one possible_ solution might be to add it as an option:

```python
class Llvm(ConanFile):

    options = {
        "enable_debug": [True, False],
    }
    default_options = {
        "enable_debug": False
    }

    def validate(self):
        ...
        if self.settings.build_type == "Debug" and not self.options.enable_debug:
            raise ConanInvalidConfiguration("Debug builds are too heavy for CCI") 
```

I don't think the impact of triggering a different package id here would be a deal breaker - this would happen when specifying a debug build type anyway. The question i guess is what should happen if build_type=Release and enable_debug=True? I suppose this could be invalidated in `validate()`. Parallel builds as an option is slightly more tricky - you wouldn't want this to affect the package id at all. Maybe an environment variable? Seems a bit messy though.

ruilvo commented on recipes/llvm/all/conandata.yml:
```diff
@@ -0,0 +1,4 @@
+sources:
+  13.0.0:
```
Can we add 13.0.1 and 14.0 (already released)?
https://github.com/llvm/llvm-project/releases

rdeterre replied:
Thanks for pointing this out, I added both releases

ruilvo commented on recipes/llvm/all/conanfile.py:
```diff
@@ -0,0 +1,128 @@
+from conans.errors import ConanInvalidConfiguration
+import os, shutil, glob
+
+projects = [
```
[1/2] Why didn't you split projects and runtimes (more on this below)?
https://llvm.org/docs/CMake.html#frequently-used-llvm-related-variables

Namely:

> LLVM_ENABLE_PROJECTS:STRING
>     Semicolon-separated list of projects to build, or all for building all (clang, lldb, lld, polly, etc) projects. This flag assumes that projects are checked out side-by-side and not nested, i.e. clang needs to be in parallel of llvm instead of nested in llvm/tools. This feature allows to have one build for only LLVM and another for clang+llvm using the same source checkout. The full list is: clang;clang-tools-extra;cross-project-tests;libc;libclc;lld;lldb;openmp;polly;pstl
> LLVM_ENABLE_RUNTIMES:STRING
>     Build libc++, libc++abi, libunwind or compiler-rt using the just-built compiler. This is the correct way to build runtimes when putting together a toolchain. It will build the builtins separately from the other runtimes to preserve correct dependency ordering. If you want to build the runtimes using a system compiler, see the [libc++ documentation](https://libcxx.llvm.org/BuildingLibcxx.html). Note: the list should not have duplicates with LLVM_ENABLE_PROJECTS. The full list is: compiler-rt;libc;libcxx;libcxxabi;libunwind;openmp To enable all of them, use: LLVM_ENABLE_RUNTIMES=all

ruilvo replied:
For some context, I was just building LLVM, and this is what I used:
```sh
cmake -S llvm -B build -G Ninja -DLLVM_ENABLE_PROJECTS='clang;clang-tools-extra;libclc;lld;lldb;polly;pstl' -DLLVM_ENABLE_RUNTIMES='libcxx;openmp'  -DLLVM_PARALLEL_LINK_JOBS=2 -DLLVM_PARALLEL_COMPILE_JOBS=8 -DCMAKE_BUILD_TYPE=Release  -DCMAKE_INSTALL_PREFIX=D:\Programs\LLVM_trunk 
```

rdeterre replied:
Thanks for this. I brought your suggested changes in.

ruilvo commented on recipes/llvm/all/conanfile.py:
```diff
@@ -0,0 +1,128 @@
+        cmake = CMake(self, parallel=False);
+        cmake.configure(
+            defs = {
+                'LLVM_ENABLE_PROJECTS': ';'.join(enabled_projects),
```
[2/2] Here:
You're adding everything as projects, and things in "compiler-rt;libc;libcxx;libcxxabi;libunwind;openmp", should be added with LLVM_ENABLE_RUNTIMES. 

Anything that's added as a RUNTIME should *not* be added as a PROJECT. 

Also, compiler-rt;libc;libcxxabi;libunwind; won't build with MSVC
cross-project-tests neither, and shouldn't be added.

ruilvo replied:
```suggestion
                'LLVM_ENABLE_PROJECTS': ';'.join(enabled_projects),
                'LLVM_ENABLE_RUNTIMES': ';'.join(enabled_runtimes),
```

ruilvo commented on recipes/llvm/all/conanfile.py:
```diff
@@ -0,0 +1,128 @@
+        if self.settings.compiler == "gcc" and tools.Version(self.settings.compiler.version) < "10":
+            raise ConanInvalidConfiguration("Compiler version too low for this package.")
+
+        if self.settings.compiler == "Visual Studio" and Version(self.settings.compiler.version) < "16.4":
```
Now, to get ready Conan 2.0, keep in mind that "msvc" is the "new" name for VS compilers. 
Use `(self.settings.compiler == "Visual Studio" or self.settings.compiler == "msvc")`

rdeterre replied:
Thanks, I changed this

ruilvo commented on recipes/llvm/all/test_package/test_package.cpp:
```diff
@@ -0,0 +1,452 @@
+// Kaleidoscope example from https://llvm.org/docs/tutorial/MyFirstLanguageFrontend/LangImpl02.html#full-code-listing
```
This is kind-of a complex test/example, I hope it build with LLVM 14.0 as well.

ruilvo commented on recipes/llvm/config.yml:
```diff
@@ -0,0 +1,3 @@
+versions:
+  13.0.0:
```
As before, please add 13.0.1 and 14.0.0.

ruilvo reviewed (changes requested):
Needs some changes.

ruilvo commented:
Btw, on the memory discussion: I'm building LLVM trunk while I write this comment. Release build, Windows system. Memory is under 2gb. 
LLVM's cmake have these options:
`LLVM_PARALLEL_COMPILE_JOBS` and `LLVM_PARALLEL_LINK_JOBS`.
Since linking is *very* notorious for it's RAM usage I suggest using `-DLLVM_PARALLEL_LINK_JOBS=2` as a CMake argument, at least on Debug builds.

ruilvo commented on recipes/llvm/all/conanfile.py:
```diff
@@ -0,0 +1,128 @@
+from conans.errors import ConanInvalidConfiguration
+import os, shutil, glob
+
+projects = [
+    'clang',
+    'clang-tools-extra',
+    'compiler-rt',
+    'debuginfo-tests',
+    'libc',
+    'libclc',
+    'libcxx',
+    'libcxxabi',
+    'libunwind',
+    'lld',
+    'lldb',
+    'mlir',
+    'openmp',
+    'parallel-libs',
+    'polly',
+    'pstl'
+]
+
+default_projects = [
+    'clang',
+    'compiler-rt'
+]
```
```suggestion
# https://llvm.org/docs/CMake.html#frequently-used-llvm-related-variables
projects = [
    'clang',
    'clang-tools-extra',
    # 'cross-project-tests', # (ruilvo) Not sure about this one
    'libc',
    'libclc',
    'lld',
    'lldb',
    'openmp',
    'polly',
    'pstl',
]

runtimes = [
    'compiler-rt',
    'libc',
    'libcxx',
    'libcxxabi',
    'libunwind',
    'openmp',
]

default_projects = [
    'clang',
]

default_runtimes = [
    'libcxx',
]
```

ruilvo commented on recipes/llvm/all/conanfile.py:
```diff
@@ -0,0 +1,128 @@
+    no_copy_source = True
+    _source_subfolder = 'source_subfolder'
+
+    options = {**{ 'with_' + project : [True, False] for project in projects }, **{
+        'fPIC': [True, False]
+    }}
+    default_options = {**{ 'with_' + project : project in default_projects for project in projects }, **{
+        'fPIC': True
+    }}
```
```suggestion
    options = {
        **{"with_project_" + project: [True, False] for project in projects},
        **{"with_runtime_" + runtime: [True, False] for runtime in runtimes},
        **{"fPIC": [True, False]},
    }

    default_options = {
        **{"with_project_" + project: project in default_projects for project in projects},
        **{"with_project_" + runtime: runtime in default_runtimes for runtime in runtimes},
        **{"fPIC": True},
    }
```

ruilvo commented on recipes/llvm/all/conanfile.py:
```diff
@@ -0,0 +1,128 @@
+            tools.check_min_cppstd(self, '14')
+
+    def _cmake_configure(self):
+        enabled_projects = [project for project in projects if getattr(self.options, 'with_' + project)]
```
```suggestion
        enabled_projects = [project for project in projects if getattr(self.options, 'with_project_' + project)]
        enabled_runtimes = [runtime for runtime in projects if getattr(self.options, 'with_runtime_' + project)]
```

ruilvo commented on recipes/llvm/all/conanfile.py:
```diff
@@ -0,0 +1,128 @@
+
+    def _cmake_configure(self):
+        enabled_projects = [project for project in projects if getattr(self.options, 'with_' + project)]
+        self.output.info('Enabled LLVM subprojects: {}'.format(', '.join(enabled_projects)))
```
```suggestion
```

ruilvo replied:
Or add both prints, if that's what you want...

ruilvo commented on recipes/llvm/all/conanfile.py:
```diff
@@ -0,0 +1,128 @@
+
+        if self.settings.build_type == "Debug":
+            raise ConanInvalidConfiguration("Debug builds are too heavy for CCI")
+
```
We'll need to add validations to **not** have something as a project and as a runtime at the same time.
And to avoid having MSVC try to build what it can't (I'm still trying to figure out what's that exactly).

ruilvo replied:
From my tests, compiler_rt can build as a RUNTIME on MSVC. libc can't build as a runtime. Haven't tested as a PROJECT, but I doubt it. (It seems to depend on the target "libc_benchmarks" that can't be built... It can probably be fixed upstream, but nevermind that...)

ruilvo replied:
> CMake Error at D:/User/Repos/llvm-project/libc/benchmarks/CMakeLists.txt:55 (add_dependencies):
>   The dependency target "benchmark" of target "libc-benchmark" does not
>   exist.
When trying as a RUNTIME

ruilvo replied:
`libc` seems to build as a PROJECT on MSVC. wow!

ruilvo commented:
I added #9948 as a way to share my ideas. It's a WIP.

rdeterre committed:
Add RTTI option

rdeterre committed:
Add enable_debug option

rdeterre committed:
Add LLVM 13.0.1 and 14.0.0

rdeterre committed:
Separate projects and runtimes

rdeterre committed:
Visual Studio or msvc

alazartech committed:
Add libcxxabi

alazartech committed:
Use options from llvm-core

alazartech committed:
Fixes

alazartech committed:
fPIC management

alazartech committed:
short_paths = True

jusito commented:
I took a look at job pool calculation, iam still unsure if this is solid enough to add it, but I want to at least share it because I brought it up.

1. Noting `LLVM_PARALLEL_COMPILE_JOBS` was quite helpful, assuming we know how much ram per compile / link job is consumed.
   - We could calculate it in python but this needs another dependency for cross platform
   - We could patch HandleLLVMOptions.cmake to calculate it from within cmake (patch below)
2. One issue is that link and compile jobs can run at the same time and that there is no way to say "this job pool is exclusivly" or a ninja option for available memory like "-l x" for cpu.
   - Logged the process history and used clang-12 + ld (not lld), seems like max(link job ram) + 2*max(compile job ram) should be enough
   - used: `cmake -G Ninja -S llvm -B build -DLLVM_ENABLE_PROJECTS='clang;clang-tools-extra;libc;libclc;lld;lldb;openmp;polly;pstl;compiler-rt;libc;openmp;' -DLLVM_ENABLE_RUNTIMES='libcxx;libcxxabi;libunwind;'`
3. another issue is that there are only certain jobs which are needing lots of memory, libLLVM.so, clang-tidy, ... so actually most optimized way would be to have individual pools per target, or maybe run it multiple times with different targets/config. Which I dont think is a good idea to add here.
4. The resource usage and runtime, and therefore overlapping resource usage, can change a lot according to used compiler, linker, architecture, ninja version e.g. the new heuristics for ninja in pr 2019 also explained in ninja issue 1335.

The easiest way is therefore to set `LLVM_PARALLEL_LINK_JOBS=1`, not using ninja or limit ninja jobs.
A more optimized solution is this patch, giving it a high value for `LLVM_RAM_PER_LINK_JOB` will have the same effect on system with less memory but allows greater machine to use some of the available resources. Some compile jobs take about 2gb, link about 10gb and the spike is around 1 link + 2 compile jobs => 14gb for link to ensure there is enough ram during both. To be safe maybe add a bit more, just to be stable.

```patch
diff --git a/llvm/cmake/modules/HandleLLVMOptions.cmake b/llvm/cmake/modules/HandleLLVMOptions.cmake
index fcaa8f20bf94..00c6fbf6fa8e 100644
--- a/llvm/cmake/modules/HandleLLVMOptions.cmake
+++ b/llvm/cmake/modules/HandleLLVMOptions.cmake
@@ -31,8 +31,25 @@ string(TOUPPER "${LLVM_ENABLE_LTO}" uppercase_LLVM_ENABLE_LTO)
 
 # Ninja Job Pool support
 # The following only works with the Ninja generator in CMake >= 3.0.
+if (NOT LLVM_RAM_PER_COMPILE_JOB)
+  set(LLVM_RAM_PER_COMPILE_JOB "2000")
+endif()
 set(LLVM_PARALLEL_COMPILE_JOBS "" CACHE STRING
   "Define the maximum number of concurrent compilation jobs (Ninja only).")
+
+cmake_host_system_information(RESULT AVAILABLE_PHYSICAL_MEMORY QUERY AVAILABLE_PHYSICAL_MEMORY)
+cmake_host_system_information(RESULT NUMBER_OF_LOGICAL_CORES QUERY NUMBER_OF_LOGICAL_CORES)
+if(LLVM_RAM_PER_COMPILE_JOB)
+  math(EXPR memory_available_jobs "${AVAILABLE_PHYSICAL_MEMORY} / ${LLVM_RAM_PER_COMPILE_JOB}" OUTPUT_FORMAT DECIMAL)
+  if (memory_available_jobs LESS 1)
+    set(memory_available_jobs 1)
+  endif()
+  if (memory_available_jobs LESS NUMBER_OF_LOGICAL_CORES)
+    set(LLVM_PARALLEL_COMPILE_JOBS "${memory_available_jobs}")
+  else()
+    set(LLVM_PARALLEL_COMPILE_JOBS "${NUMBER_OF_LOGICAL_CORES}")
+  endif()
+endif()
 if(LLVM_PARALLEL_COMPILE_JOBS)
   if(NOT CMAKE_GENERATOR MATCHES "Ninja")
     message(WARNING "Job pooling is only available with Ninja generators.")
@@ -42,8 +59,22 @@ if(LLVM_PARALLEL_COMPILE_JOBS)
   endif()
 endif()
 
+if (NOT LLVM_RAM_PER_LINK_JOB)
+  set(LLVM_RAM_PER_LINK_JOB "14000")
+endif()
 set(LLVM_PARALLEL_LINK_JOBS "" CACHE STRING
   "Define the maximum number of concurrent link jobs (Ninja only).")
+if(LLVM_RAM_PER_LINK_JOB)
+  math(EXPR memory_available_jobs "${AVAILABLE_PHYSICAL_MEMORY} / ${LLVM_RAM_PER_LINK_JOB}" OUTPUT_FORMAT DECIMAL)
+  if (memory_available_jobs LESS 1)
+    set(memory_available_jobs 1)
+  endif()
+  if (memory_available_jobs LESS NUMBER_OF_LOGICAL_CORES)
+    set(LLVM_PARALLEL_LINK_JOBS "${memory_available_jobs}")
+  else()
+    set(LLVM_PARALLEL_LINK_JOBS "${NUMBER_OF_LOGICAL_CORES}")
+  endif()
+endif()
 if(CMAKE_GENERATOR MATCHES "Ninja")
   if(NOT LLVM_PARALLEL_LINK_JOBS AND uppercase_LLVM_ENABLE_LTO STREQUAL "THIN")
     message(STATUS "ThinLTO provides its own parallel linking - limiting parallel link jobs to 2.")

```

jusito commented:
For anyone interested, update the recipe but wasnt sure to create a pr to conan-center or to [rdeterres work](https://github.com/rdeterre/conan-center-index/pull/1).

- Fixed the issue with component.json not generated
- merged the recipe with llvm-core
- used ninja an job calculation
- fixed openmp / libclang not generated or .so only in static builds and vice versa
- added an regex option to keep matching binaries (very handy for some projects)

Only issue left for shared build, test is failing because of:
```
: CommandLine Error: Option 'use-dbg-addr' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options
Aborted (core dumped)
```

Only this part i have no idea how to fix it or if its needed:
```python3
self.cpp_info.components[component].build_modules["cmake_find_package"].extend([
    self._alias_module_file_rel_path,
    self._old_alias_module_file_rel_path,
])
self.cpp_info.components[component].build_modules["cmake_find_package_multi"].extend([
    self._alias_module_file_rel_path,
    self._old_alias_module_file_rel_path,
])
```

tcanabrava commented:
I Kindly request help from the conan team here. 

LLVM is a really important piece of stack and we are trying to merge this request since at least LLVM 9. I took over an abandoned MR and ported to a newer LLVM version, then abandoned if after a while, someone took over, ported to another newer LLVM version, then abandoned it after a while, then someone else took over...

I'm currently using different bash / powershell scripts to download / compile llvm with the flags that I need on all systems bypassing conan, and that's really not something that I should be worried about doing.

Should we keep pushing this MR, should we wait for conan2, what should we do here?
@rdeterre , @Manu343726 , @jusito and all?

ruilvo commented:
@tcanabrava 

Fine you say that... I gave up on trying to package llvm/clang for Conan exactly because it doesn't support a workflow that I'd like... I'd like for it to be possible for Conan to download or build clang, and then use that just built compiler to all the other dependencies and the project itself... But that can't be done... So unless you're using the llvm libraries, packaging compilers isn't useful.

tcanabrava commented:
> @tcanabrava
> 
> Fun you say that... I gave up on trying to package llvm/clang for Conan exactly because it doesn't support a workflow that I'd like... I'd like for it to be possible for Conan to download or build clang, and then use that just built compiler to all the other dependencies and the project itself... But that can't be done... So unless you're using the llvm libraries, packaging compilers isn't useful.

That's exactly what I'm doing, using libclang and libllvm. I'm not trying to use the binaries as compilers.

jusito commented:
I would be happy to see llvm in public conan. I had no time for a look into conan 2, so I have no idea how long it will take.
Which state of llvm recipe are you using right now?

We are using the version from [my pr](https://github.com/rdeterre/conan-center-index/pull/1) daily, updated it for 15.0.0. There you have the possibility to keep certain binaries clang / clang++ aso which we are using to decouple our application from a specific compiler. We are resolving the current llvm binary folder from conan and using clang to compile test files to llvm IR. So I can work with clang 15 but the project is using clang-14 for IR file generation. Secondly cmake can be used from conan with `conan install cmake/3.23.3@ -g virtualrunenv`, so using clang in the same way seems possible for me, I just have no idea how to support virtualrunenv generator. If you want to compile with clang, maybe including conan once for llvm/clang dep, extracting path, setting compiler, rerun autodetect + add other dependencies second? 

Ok tested it with my state of the recipe: @ruilvo 
1. exported from pr to llvm/14.0.6@phasar/develop
2. `conan install llvm/14.0.6@phasar/develop -s build_type=Release -o llvm:keep_binaries_regex="^(clang|clang\+\+|opt)$" -g virtualrunenv`
3. conan profile shouldn't set CC/CXX
4. your environment CC/CXX should point to clang/clang++
5. compile your application, clang 14.0.6 from conan is used instead of 15.0.0, even if conan profile is set to 15.

bmanga commented:
I too would like to see an llvm recipe in conan

jusito commented:
Conan 2.0 is out to continue here how about collection some requirements?

I planned to invest some time into llvm recipe until end of april, I have following issues on my list:
- conan 2.0 check
- for compiler-rt there are additional headers on the system needed, e.g. on ubuntu x64 package libc6-dev-i386 is needed, how to check for that?
- compatibility settings, why rebuild llvm without clang if there is a package which contains clang?
- Currently if consumed transitive llvm dependencies aren't resolved so need to add transitive stuff to link targets, not sure if its a recipe or a consuming issue.
- cleanup the recipe, some methods are quite huge
- probably need to create an own mr for llvm into conan-center-index instead of rdeterre

It would help me a lot if:
- I get feedback regarding the pr into rdeterres work e.g. a review or a test if it works in other environments
- additional requirements or help regarding the list above
- other suggestions how to continue

cheers

ericriff commented on recipes/llvm/all/conanfile.py:
```diff
@@ -0,0 +1,329 @@
+            'LLVM_ENABLE_LIBPFM': False,
+            'LLVM_ENABLE_LIBEDIT': False,
+            'LLVM_ENABLE_FFI': self.options.with_ffi,
+            'LLVM_ENABLE_ZLIB': self.options.get_safe('with_zlib', False),
```
We should set these variables to `FORCE_ON` instead of True so they add the `REQUIRED` keyword. Otherwise if CMake fails to find the dependencies they get silently ignored.

ericriff replied:
This is the code form version 16.0.03
```cmake
if(LLVM_ENABLE_ZLIB)
  if(LLVM_ENABLE_ZLIB STREQUAL FORCE_ON)
    find_package(ZLIB REQUIRED)
  elseif(NOT LLVM_USE_SANITIZER MATCHES "Memory.*")
    find_package(ZLIB)
  endif()
  if(ZLIB_FOUND)
  ...
```

jusito replied:
Its also there in in llvm 13 and documentation showed that this is the case for other options as well. Can add it in my pr.

jcar87 commented:
We are currently not prioritising packages that add toolchains (rather than libraries or pure executables), as we are prioritising other issues in our backlog.

LLVM itself has a myriad of components, and I suspect different users would have different needs. We are opening this issue to track this: https://github.com/conan-io/conan-center-index/issues/23834  - actual uses cases with examples will help the team gain better insights at how a potential llvm recipe may be used, whether or not it is worth packaging different LLVM components in separate recipes, and so on. Any information of this would be appreciated in that issue so that we can move forward, thanks!
